sixtyfps_runtime/rendering_backends/qt/qt_window.rs

```diff
@@ -402,16 +402,65 @@ impl ItemRenderer for QtItemRenderer<'_> {
             painter->setFont(font);
             painter->setPen(QPen(fill_brush, 0));
             painter->setBrush(Qt::NoBrush);
             if (!elide) {
                 painter->drawText(rect, flags, string);
+            } else if (!(flags & Qt::TextWordWrap)) {
+                QString elided;
+                QFontMetrics fm(font);
+                while (!string.isEmpty()) {
+                    int pos = string.indexOf('\n');
+                    if (pos < 0) {
+                        elided += fm.elidedText(string, Qt::ElideRight, rect.width());
+                        break;
+                    }
+                    QString line = string.left(pos);
+                    elided += fm.elidedText(line, Qt::ElideRight, rect.width());
+                    elided += '\n';
+                    string = string.mid(pos + 1);
+                }
+                painter->drawText(rect, flags, elided);
             } else {
-                auto elided = QFontMetrics(font).elidedText(string, Qt::ElideRight, rect.width());
+                // elide and word wrap: we need to add the elipsis manually on the last line
+                QString elided = string;
+                QFontMetrics fm(font);
+                QTextLayout layout(string, font);
+                QTextOption options;
+                options.setWrapMode(QTextOption::WordWrap);
+                layout.setTextOption(options);
+                layout.setCacheEnabled(true);
+                layout.beginLayout();
+                int leading = fm.leading();
+                qreal height = 0;
+                int last_line_begin = 0, last_line_size = 0;
+                while (true) {
+                    auto line = layout.createLine();
+                    if (!line.isValid()) {
+                        last_line_begin = string.size();
+                        break;
+                    }
+                    line.setLineWidth(rect.width());
+                    auto lf = QStringView(string).mid(line.textStart(), line.textLength()).indexOf('\n');
+                    if (lf >= 0) {
+                        line.setNumColumns(lf);
+                    }
+                    height += leading + line.height();
+                    if (height > rect.height()) {
+                        break;
+                    }
+                    last_line_begin = line.textStart();
+                    last_line_size = line.textLength();
+                }
+                if (last_line_begin < string.size()) {
+                    elided = string.left(last_line_begin);
+                    QString to_elide = QStringView(string).mid(last_line_begin, last_line_size).trimmed() % QStringView(u"…");
+                    elided += fm.elidedText(to_elide, Qt::ElideRight, rect.width());
+                }
                 painter->drawText(rect, flags, elided);
```

Comment:
This will do the entire work of line breaking a second time though. I think it would be more efficient to call `layout.endLayout();` and then draw it using `QTextLayout::draw()`, followed by one more call for the elision character if needed. The x position of that would I think be `QTextLine::naturalTextWidth()` of the last line.

Reply:
true, but knowing exactly where to cut the last line and put the elipsis is a bit of work. I'll keep it like this for now.